crates/wasmtime/src/runtime/vm/vmcontext.rs

```diff
@@ -623,6 +660,38 @@ mod test_vmshared_type_index {
     }
 }
 
+/// A WebAssembly tag defined within the instance.
+///
+#[derive(Debug)]
+#[repr(C, align(8))]
+pub struct VMTagDefinition {
+    /// Function signature's type id.
+    pub type_index: VMSharedTypeIndex,
+}
```

Comment:
Some questions about this type:
- Why insist on it being 8-byte aligned? For our purposes in `TaggedPointer`, the natural 4 byte alignment already gives us the required 2 bits.
- The `type_index` field seems to never actually be used. On the other hand, I'm not necessarily opposed to storing it here, given that the type needs to have a size of at least 4 bytes anyway for the `TaggedPointer` use.

In any case, we should at least have a comment here to indicate that there's a need to keep the alignment of this type in sync with what `TaggedPointer` expects.

Reply:
The `type_index` is necessary for the embedder API -- which is unimplemented atm. So the contents are irrelevant as long as the structure is aligned. We only use the runtime address.

Reply:
If I lower the size then I get a misaligned pointer dereference on my machine. Maybe it should be `align(16)` like globals?

Reply:
Did you change the `align` directive on `VMTagDefinition` without updating `size_of_vmtag_definition` in the `VMOffsets`?

Reply:
I have a 64 bit machine, so the alignment must be a multiple of `0x8`, which `0x4` certainly isn't.

Reply:
I have changed the alignment to 16.

Reply:
Are you saying that on a 64 bit machine each struct must be 8-byte aligned? That's simply wrong. After all, if you remove the `align` from the definition of `VMTagDefinition` altogether, you get that `VMTagDefinition` an alignment of 4 bytes, as expected.

The reason why you were seeing failures (and why changing the alignment to 16 fixes it) is simply the layout within the `VMContext` fields: If you look at the usage of the `fields!` macro where the fields get defined, we put `defined_tags` before fields like `defined_func_refs`, `typed_continuations_stack_chain`. These fields need an 8 byte alignment, which gets broken if `defined_tags` appears beforehand without re-establishing the required alignment for the subsequent fields. Previously this worked because of the `align(16)` before `defined_globals`.

So if we change the size/alignment of `VMTagDefinition` to 4 bytes, you either need to move the `defined_tags` between `owned_memories` and `defined_globals`  in the `fields!`, or put an `align(8)` behind the definition of  `defined_tags` .

Reply:
> Are you saying that on a 64 bit machine each struct must be 8-byte aligned? That's simply wrong. After all, if you remove the `align` from the definition of `VMTagDefinition` altogether, you get that `VMTagDefinition` an alignment of 4 bytes, as expected.

No that's not what I am saying. What I am saying is that we need to ensure that they are laid out in the vmctx such that they aligned, otherwise we can obtain an unaligned pointer.

Reply:
It makes no different whether we make `VMTagDefinition` 4 bytes wide or add an alignment modifier to it. The instances of `VMTagDefinition` lives only in the vmctx. It is simpler to put an alignment modifier, otherwise the offset calculation will have to take into account that `size_of_vmtag_definition()` needs + alignment. The key invariant we need to preserve is that the `VMTagDefinition`s are laid out in the vmctx, such that a pointer to a `VMTagDefinition` base is aligned.

Reply:
Glad we are on the same page about that part.

Can we now please achieve the correct alignment of fields within the `VMContext` in the usual way and without unnecessarily increasing the size/alignment of every single `VMTagDefinition`? To do that, we can simply move the `defined_tags` behind the `defined_memories`, or add an `align` after the `defined_tags`.

Reply:
> It makes no different whether we make `VMTagDefinition` 4 bytes wide or add an alignment modifier to it. 

The problem you were seeing when decreasing the alignment of `VMTagDefinition` is not due to the tag definitions in the `VMContext` themselves being misaligned, but the fields in the `VMContext` *after* the tag definitions, because those require a larger alignment than `VMTagDefinition`'s 4 bytes, but that was no longer reflected in the `fields!`. That's why I suggested adding an alignment change to 8 *after* `defined_tags` in the `fields!`, not adding one immediately before the `defined_tags`. Or just moving the `defined_tags` right before the existing `align(16)` for `defined_globals`.

Reply:
Here's the concrete change I'm suggesting:
https://github.com/frank-emrich/wasmtime/commit/5d8c6b24488f793aba9a699ae21d7f9a5d9d4ea2

Reply:
To illustrate my point see https://godbolt.org/z/dEczKo34T

Reply:
> Here's the concrete change I'm suggesting: [frank-emrich@5d8c6b2](https://github.com/frank-emrich/wasmtime/commit/5d8c6b24488f793aba9a699ae21d7f9a5d9d4ea2)

But now from fields in `VMTagImport` can have unaligned pointers. We need to align the elements in the array.

Reply:
> To illustrate my point see https://godbolt.org/z/dEczKo34T

Your `is_aligned` function is simply wrong. Why are you requiring that a pointer to `tag_definition` must be divisible by 8? All that is required is for it to be divisible by 4, since `_Alignof(struct tag_definition)` is 4 bytes.

The correct version of is_aligned for `tag_definition` would look like this:

```
bool is_aligned(const struct tag_definition* ptr) {
  return ((uintptr_t) ptr) % _Alignof(struct tag_definition) == 0;
}
```

In your type `tag_definition_aligned` you are not making sure that the type is "correctly" aligned (it is so automatically), but you are asking for it to be 16-byte aligned. This then makes it pass your own (faulty) alignment test.

Reply:
No you are misunderstanding. The alignment of a x64 is 8 bytes, so the bottom three bits must be zero, hence 0x7.

Reply:
You seem to forget that we are talking about pointers.

Reply:
Let's say that we make `VMTagDefinition` 4 bytes wide. For simplicity lets assume allocation of the `defined_tags` array starts at 0x0, then we have
```
+-----+-----+-----+-----+
| 0x0 | 0x4 | 0x8 | 0xC |
+-----+-----+-----+-----+
^--p0 ^--p1 ^--p2 ^--p3
```
Then pointers p0-p3 are aligned on a 32 bit platform, but on a 64 bit platform only p0 and p2 are aligned.

Reply:
I read your picture so that
```
println!("{:p}", p0);
println!("{:p}", p1);
println!("{:p}", p2);
println!("{:p}", p3);
```

should output
```
0x0
0x4
0x8
0xC
```
Right?

In that case, all pointers are correctly aligned, even on x64.

Reply:
Maybe the confusion stems from that the pointers we care about points into the heap. We don't care about the alignment of the structure it points to, however, but padding the structure we get that any pointer to it will be properly aligned.

Reply:
```
0x0 = 0b0000
0x4 = 0b0100
0x8 = 0b1000
0xC = 0b1010
```
Clearly, the heap addresses 0x4 and 0xC are not aligned, they fail the "bottom three bits" test.

Reply:
I think you are confusing structure alignment and heap alignment.

Reply:
We seem to be in some sort of disagreement and should resolve this by showing a minimal, standalone example to someone.
Do you think that the example in  https://godbolt.org/z/dEczKo34T demonstrates that the pointers `&ts[1]` and `&ts[3]` are misaligned?

Reply:
Actually, we should never need to dereference the pointer, so it shouldn't matter whether it is aligned (we only use it for an identity check). Then we can pack the `VMTagDefinition` into a dense array with element size 4 bytes. But it will be a problem if someone dereferences a `from` field in `VMTagImport`, I think.

Reply:
> We seem to be in some sort of disagreement and should resolve this by showing a minimal, standalone example to someone. Do you think that the example in https://godbolt.org/z/dEczKo34T demonstrates that the pointers `&ts[1]` and `&ts[3]` are misaligned?

You are confusing structure alignment with heap alignment. The pointers `&ts[1]` and `&ts[3]` are not heap aligned (they are structure aligned; I am abusing the stack, but you get the same results with `malloc/free`).

Reply:
Could you update that example to illustrate that?

Reply:
Just malloc the arrays instead of stack allocating them.

If you think about the theory of pointer tagging, then it is clear that the bottom bits must always be zero on a properly aligned heap pointer (not necessarily true for a pointer into the stack!). I can recommend this primer https://dl.acm.org/doi/pdf/10.1145/3558200.

Reply:
My latest commit makes the `VMTagDefinition` 4 byte **structure** aligned. It works because we never dereference any pointer to it.

Reply:
I think I understand where the confusion is coming from. I was talking about heap aligned pointers, because I had the content of tagged pointers paged in. But you were talking about pointers to some structure. We never treat the pointer as a heap pointer, so it is ok to pack the arrays.

Reply:
Let's ignore pointer tagging for a second. Do you believe that in the following program, without any additional context, there is any kind of problem with alignment, simply in the sense of violating C alignment rules? (Or any other kind of violation of the C standard?)
https://godbolt.org/z/6sdhcnW73

Coming back to pointer tagging: In our particular case, we only use the lowest 2 bits for tagging, the rest must remain untouched. If we ever changed that and wanted to increase the number of bits we want to tag, we would indeed need to impose additional alignment requirements. I don't think there was ever any disagreement on that part.

Reply:
No. I was talking about aligned heap pointers. A pointer is to a heap object is only aligned if its bottom bits (platform specific) are zero.

Reply:
I don't think pointer alignment is part of the C standard. I was talking about what it means for a heap pointer to be properly aligned.

Reply:
I would like to dig into what you mean by "heap aligned pointers". You seem to use this term as if it was some well-defined requirement on certain pointers, distinct from the alignment requirements on structs (or values in general). The following question will hopefully help us get on the same page.

1. From what I can tell, this would be your definition on x64: "A pointer `ptr` is heap aligned iff it points to an object on the heap and we have ptr % 8 == 0". Correct?
2. Who imposes that certain pointers have to be heap aligned? The ISA, the OS, or a particular programming language? 
3. Following the previous question, your notion of heap aligned pointers seems to be independent from us wanting to tag certain pointers. Of course you can only tag pointers if they satisfy certain alignment requirements (so that there are guaranteed trailing 0), but you seem to say that certain (or all) pointers must be heap aligned even in the absence of pointer tagging. Correct?
5. What pointers must be heap aligned? All of them? (Of course, the definition above does not impose anything on pointers into the stack).
6. Using the definition in 1., and assuming that *all* pointers must be correctly heap aligned, my example at https://godbolt.org/z/6sdhcnW73 would violate that, because `&a[1] % 8 != 0`. Do you agree?
7. Regarding the TODO you added on the `from` field in `VMTagImport` in c2c9e669e1e3fec2dc781394ac4614db89254f78. What would happen if we were to read from that pointer? Would it simply be UB (according to Rust, the ISA, OS, ...) to read from that pointer, but we may still read a valid `VMTagDefinition` (i.e., because we previously stored a `VMTagDefinition` at that pointer)? Or would we be reading garbage, because we didn't previously store a `VMTagDefinition` at that location?

Reply:
We are still talking past each other, partly because you were thinking about the alignment of a pointer into a structure, and I was thinking about base pointers to heap allocated objects. For the latter, the lower N bits must be all zero, that's how the ABI defines it.

1. No. It must be `ptr % 16`, because it must be aligned to the largest element of the architecture -- so I was wrong when I said 8 before.
2. ABI/malloc.
3. No. Pointer tagging only works because of the how ABI/malloc/free works. If addresses on the heap were allocated byte-wise, then you wouldn't have any free bits.
4. Pointers returned by malloc.
5. No.
6. Dereferencing a pointer in Rust is only OK as long as it is properly align. I thought about it, and the pointer is properly aligned to the structure, which is what matters here. But it may not be a properly aligned base pointer to a heap allocated object; but it is an aligned pointer to an element of an aggregate object on the heap. 

The tagged pointer is crucial to the context here, because the tags are passed around as tagged pointers. Currently, it works because the lower 2 bits are always zero, but this would no longer be true if we made the VMTagDefinition byte-sized, so we have to be careful. The access into the aggregate structure is aligned, but interpreting such a pointer as a base pointer to a heap allocated object is not aligned in the sense of malloc/free.

I agree that your suggestion works and I was wrong to think that the elements of aggregate had to be aligned as per malloc, because we do not need quite that. Currently, we only need that 4 byte alignment.

Reply:
> Currently, it works because the lower 2 bits are always zero, but this would no longer be true if we made the VMTagDefinition byte-sized, so we have to be careful. The access into the aggregate structure is aligned, but interpreting such a pointer as a base pointer to a heap allocated object is not aligned in the sense of malloc/free.

Yes, I agree that things would absolutely break if the alignment of `VMTagDefinition` became smaller than 4 bytes. That's why I suggested the following in my very first comment in this thread: "In any case, we should at least have a comment here to indicate that there's a need to keep the alignment of this type in sync with what TaggedPointer expects."


> I was thinking about base pointers to heap allocated objects. For the latter, the lower N bits must be all zero, that's how the ABI defines it.

>  It must be ptr % 16, because it must be aligned to the largest element of the architecture -- so I was wrong when I said 8 before.


This is the part where seem to disagree. Luckily, this disagreement does not seem to affect anything in this PR anymore, but for the sake of resolving this dispute:

`malloc` indeed returns a pointer that's guaranteed to have the largest alignment on the current platform, which is 16 bytes on the platforms we care about. We both agree on that.
But you seem to be under the impression that whenever you want to store a standalone struct on the heap (where "standalone" means not contained in another struct, not part of an array of such structs, ...), it must be a) at the beginning of an allocation returned by malloc, or b) at an address that's 16-byte aligned (I'm not sure if you're claiming version a) or b), but of course satisfying a) implies satisying b) ). I claim that this rule does not exist in this form. 

Of course, if you do `void* p = malloc(100)` and want to store a struct `S` with `sizeof(S) == 4 == _Alignof(S)` somewhere in that allocation other than at `p` itself, you have to make sure that the following holds:
- you are only allowed to store it at `p + 4*N` for some N to satify the alignment requirements of `S`
- the pointer you eventually call `free` on must of course be `p`, the beginning of the allocation originally returned by malloc

In other words, I claim that the following program is perfectly fine, and does not violate any rules imposed by C, or any ABI rules on Linux:
https://godbolt.org/z/Mx6a16fWc


Of course, in these situations you need to keep the base pointer of the alloction (i.e, pointing to the beginning of the allocation) around for the purpose of memory management so that you can eventually call `free` on the beginning of the allocation. But that's independent from where you may store objects inside that allocation in the meantime, as long as you follow the rules I mentioned above.
This seems to be where your notion of "heap allocated object" comes from: If you *do* put an object at the base pointer of an allocation returned by malloc and `free` that pointer later, you automatically satisfy that requirement, because the pointer to that object is the base pointer of the allocation. But there's no requirement that the base pointer to an allocation must also be a base pointer to some object, as shown in my godbolt link above.

Reply:
In case you complain that in my godbolt example above, the struct only contains a single element: Here's a variation where it contains two `int32_t` and we have `sizeof(S2) == 8`, but still `_Alignof(S2) == 4`. Therefore, everything still works and this does not violate any rules either:

https://godbolt.org/z/9689Wq9zz

Reply:
You are still talking past me or I am talking past you. The x64 Linux ABI dictates that the stack&heap are aligned 16 bytes on the boundary. This means that any allocation you perform, the last digit in hex notation will be 0, or in binary format, the bottom bits will be all zero. In your example above `data_ptr` is not aligned on the boundary. Indeed, calling free on it is bad. The pointer I am concerned with is the `alloc`. In my previous example I was using an array as a heap/arena to illustrate densely packed allocations and that pointers **into** to that aggregate structure can yield unaligned pointers on the boundary. Why is this important? It is important because we rely the alignment of heap allocated pointers for correctness. But I was overthinking it, because we do not rely on x64 alignment semantics, but rather just x86 semantics.

Reply:
> In case you complain that in my godbolt example above, the struct only contains a single element: Here's a variation where it contains two `int32_t` and we have `sizeof(S2) == 8`, but still `_Alignof(S2) == 4`. Therefore, everything still works and this does not violate any rules either:
> 
> https://godbolt.org/z/9689Wq9zz
> 
> (edit: and to clarify: changing the print statement to print `data_ptr->elem` instead of `data_ptr->elem2` would also be perfectly fine)

I am not disputing any of these points. It is orthogonal to what I was talking about.

Reply:
Let's put it like this: Whenever you write C (or Rust, for that matter) on x64 Linux and someone gives you a pointer `T* p` for some arbitrary type `T` and expects you to access that pointer (read or write to it), you *are* allowed to assume that the pointer is correctly aligned, which means that p % _Alignof(T) == 0 must hold. Otherwise, the alignment rules would be violated. But you *are not* allowed to assume in general that such a pointer has an even bigger alignment, in particular you cannot assume by default that p % 16 == 0. This doesn't change just because, say, `p` point to the beginning of a another struct or something like that. In other words, just because someone gives you a pointer to an object, you are not allowed to assume that it is aligned at what you call "the boundary".

The fact that the stack and allocations returned by `malloc` have the largest possible alignment enables us to maintain the rule above. In both of my godbolt examples, if `malloc` was allowed to return a pointer that's less than 4-byte aligned, there would be an alignment error (but of course it returns a 16-byte aligned pointer, so we're good).

If you want to establish a rule that in your code for certain pointers `T* p`, we have that `p` has an alignment greater than just `_Alignof(T)` bytes, this is something you as the user are doing on top of the language's or ABI's alignment rules and would have to document, because you are not allowed to assume this just be default.

Reply:
I am not arguing or disputing any of those points, I agree with you. I never meant to talk about alignment of structures. I was only ever talking about alignment of heap allocated pointers.

Reply:
Okay, I just didn't understand what problem you were previously seeing with the code, given that its correctness was established just by these standard alignment rules.

But it seems like we're on the same page now that there is no alignment problem in the current version.

Reply:
Maybe it just comes down to slightly different interpretations of the term "tagged pointer": You seem to be using a definition of tagged pointer where we assume that we only tag pointers that are maximally aligned (i.e., 4 bits usable for tagging on x64), for example because we know that they come directly from `malloc`. When I speak about tagged pointers, I assume that the number of bits you can use for tagging is inherently tied to the type `T` of data you point to, because that guarantees you that there are `log2(_Alignof(T) )` bits that you can use for tagging, but no more.